plugins/out_s3/arrow/compress.c

```diff
@@ -0,0 +1,174 @@
+ *
+ * We use GLib binding to call Arrow functions (which is implemented
+ * in C++) from Fluent Bit.
+ *
+ * https://github.com/apache/arrow/tree/master/c_glib
+ */
+
+#include <arrow-glib/arrow-glib.h>
+#include <inttypes.h>
+
+/*
+ * GArrowTable is the central structure that represents "table" (a.k.a.
+ * matrix).
+ */
+static GArrowTable* parse_json(uint8_t *json, int size)
+{
+        GArrowJSONReader *reader;
+        GArrowBuffer *buffer;
+        GArrowBufferInputStream *input;
+        GArrowJSONReadOptions *options;
+        GArrowTable *table;
+        GError *error = NULL;
+
+        buffer = garrow_buffer_new(json, size);
+        if (buffer == NULL) {
+            return NULL;
+        }
+
+        input = garrow_buffer_input_stream_new(buffer);
+        if (input == NULL) {
+            g_object_unref(buffer);
+            return NULL;
+        }
+
+        options = garrow_json_read_options_new();
+        if (options == NULL) {
+            g_print("fail to create json options: %s\n", error->message);
+            g_object_unref(buffer);
+            g_object_unref(input);
+            return NULL;
+        }
+
+        reader = garrow_json_reader_new(GARROW_INPUT_STREAM(input), options, &error);
+        if (reader == NULL) {
+            g_print("cannot create json reader: %s\n", error->message);
+            g_error_free(error);
+            g_object_unref(buffer);
+            g_object_unref(input);
+            g_object_unref(options);
+            return NULL;
+        }
+
+        table = garrow_json_reader_read(reader, &error);
+        if (table == NULL) {
+            g_print("cannot parse json: %s\n", error->message);
+            g_error_free(error);
+            g_object_unref(buffer);
+            g_object_unref(input);
+            g_object_unref(options);
+            g_object_unref(reader);
+            return NULL;
+        }
+        g_object_unref(buffer);
+        g_object_unref(input);
+        g_object_unref(options);
+        g_object_unref(reader);
+        return table;
+}
+
+static GArrowResizableBuffer* table_to_buffer(GArrowTable *table)
+{
+        GArrowSchema *schema;
+        GArrowResizableBuffer *buffer;
+        GArrowBufferOutputStream *sink;
+        GArrowRecordBatchStreamWriter *writer;
+        GError *error = NULL;
+        gboolean success;
+
+        schema = garrow_table_get_schema(table);
+        if (schema == NULL) {
+            return NULL;
+        }
+
+        buffer = garrow_resizable_buffer_new(0, &error);
+        if (buffer == NULL) {
+            g_print("cannot create buffer: %s\n", error->message);
+            g_error_free(error);
+            g_object_unref(schema);
+            return NULL;
+        }
+
+        sink = garrow_buffer_output_stream_new(buffer);
+        if (sink == NULL) {
+            g_object_unref(schema);
+            g_object_unref(buffer);
+            return NULL;
+        }
+
+        writer = garrow_record_batch_stream_writer_new(
```

Comment:
Apache Arrow has two file formats:

1. Streaming format: https://arrow.apache.org/docs/format/Columnar.html#ipc-streaming-format
2. File format: https://arrow.apache.org/docs/format/Columnar.html#ipc-file-format

This code uses the streaming format. The streaming format is suitable for stream data. Users can start processing before they receive all data. But users can't do random access.

The file format is suitable for file data. Users can't start processing before they receive all data. But they can do random access.

I think that file format is suitable for data on S3. Users can do random access by HTTP range header without downloading all data on S3.

We can use convenient `garrow_table_write_as_feather()` for the file format. Note that "Feather v2" is another name of the file format: https://arrow.apache.org/docs/python/feather.html

```c
success = garrow_table_write_as_feather(table, GARROW_OUTPUT_STREAM(sink), NULL, &errors);
```

We can compress data by LZ4 or Zstandard:

```c
GArrowFeatherWriteProperties *properties = garrow_feather_write_properties_new();
g_object_set(G_OBJECT(properties),
            "compression", GARROW_COMPRESSION_TYPE_ZSTD,
            NULL);
success = garrow_table_write_as_feather(table, GARROW_OUTPUT_STREAM(sink), properties, &errors);
g_object_unref(properties);
```

Reply:
Thank you. I changed `compress.c` to use Feather v2 format with ZSTD compression.